### DISCUSSION TITLE
Why does StackHasher hash all elements (matching filter) instead of just the one element that threw the exception?

### DISCUSSION THREAD

**Discussion by hjohn** [logfellow/logstash-logback-encoder#499] [Q&A]
I am wondering, why does the StackHasher go through all stack trace elements (filtering out some according to the exclude patterns) to compute its hash?  This seems rather expensive for little added benefit.

The reason I ask is that it seems to me that the very last stack trace element (the one where the exception is thrown) will be unique already as you can only throw a single exception from a single location and it should bubble up in the same way, so the rest of the elements will be similar, barring some proxy/cglib generated elements.

In fact, I've used this method of hashing before in large projects (hashing only the element that threw the exception), and the hash was certainly highly useful in grouping the same exceptions together without hashing a complete trace.  This also means you donot need to use AND correctly configure a filtering mechanism for unstable stack trace elements.

I suppose that "engineered" exceptions or exceptions without a stacktrace could be miss hashed, but that seems to be such a rare situation that I donot think it warrants having a full hash as the default behavior.

Maybe this simplified mechanism could be provided as an option for those unwilling to do exhaustive analysis on what exactly to include from traces to get stable hashes?

**Comment by philsttr** [comment]
Hi @hjohn 

> I am wondering, why does the StackHasher go through all stack trace elements (filtering out some according to the exclude patterns) to compute its hash? 

The stack hash implementation was a contribution from the community in #228 .  Perhaps the original author, @pismy, could explain the reasoning.

Your idea sounds pretty good to me.  Although, I think at a minimum it should hash the first stack frame of each Throwable in the cause chain.

> Maybe this simplified mechanism could be provided as an option for those unwilling to do exhaustive analysis on what exactly to include from traces to get stable hashes?

This seems reasonable to me.  Would you be willing/able to submit a PR?

**Reply by hjohn** [reply]
I can do a PR over a weekend, I'll wait a bit in case the original author wants to chime in.

I had a talk with a colleague at work, and he thought one of the reasons could be that different code paths could lead to the same exception type and location.  With the current StackHasher, these would hash differently (I presume), while a more simplistic implementation only based on the last frame would hash these the same.  Which one of these is more desirable seems to be a matter of preference.

**Reply by philsttr** [reply]
Awesome!  Thank you!

I agree that providing an option for either approach is the best solution, due to different preferences.

**Reply by pismy** [reply]
Hey @hjohn @philsttr 
Indeed my hashing algorithm looks maybe a bit overkill.

B.t.w I also developed a [similar tooling for logstash](https://github.com/elastic/logstash/issues/9333), and explained the goal and choices in a minimal [GitHub pages site](https://pismy.github.io/logstash-filter-java_stack_digest/challenge.html)
Exemple stack 3 shows the necessity of taking into account the parent error to compute an error's digest, but here your question is: **wouldn't the root cause be enought to compute a digest?**

For sure I could design a Java program where the same root cause exception might not cause the same exceptions in the calling layers, but that's maybe not the general case.

For instance think about a generic exception such as JAX RS [WebApplicationException](https://docs.oracle.com/javaee/7/api/javax/ws/rs/WebApplicationException.html): if thrown from the same call stack, the calling program may react differently/rethrow different exceptions depending on the `status` value...
